src/extlibs/hrvo/hrvo_agent.cpp

```diff
@@ -34,139 +34,202 @@
+                     float max_neighbor_dist, std::size_t maxNeighbors, float radius,
+                     float max_radius_inflation, const Vector &velocity, float maxAccel,
+                     AgentPath &path, float maxSpeed)
     : Agent(simulator, position, radius, max_radius_inflation, velocity, velocity,
             maxSpeed, maxAccel, path),
       maxNeighbors_(maxNeighbors),
-      neighborDist_(neighborDist),
-      prefSpeed_(prefSpeed),
-      uncertaintyOffset_(uncertaintyOffset)
+      max_neighbor_dist(max_neighbor_dist),
+      prefSpeed_(max_speed_ * PREF_SPEED_SCALE),
+      // TODO (#2676): This config should be dependency injected and updated when a
+      // parameter is changed
+      obstacle_factory(TbotsProto::RobotNavigationObstacleConfig()),
+      ball_obstacle(std::nullopt)
 {
 }
 
-void HRVOAgent::computeNeighbors()
+void HRVOAgent::updatePrimitive(const TbotsProto::Primitive &new_primitive,
+                                const World &world)
 {
+    AgentPath path;
+    static_obstacles.clear();
+    ball_obstacle = std::nullopt;
+    if (new_primitive.has_move())
+    {
+        const auto &motion_control = new_primitive.move().motion_control();
+        float speed_at_dest        = new_primitive.move().final_speed_m_per_s();
+        float new_max_speed        = new_primitive.move().max_speed_m_per_s();
+        setMaxSpeed(new_max_speed);
+        setPreferredSpeed(new_max_speed * PREF_SPEED_SCALE);
+
+        // TODO (#2418): Update implementation of Primitive to support
+        // multiple path points and remove this check
+        CHECK(motion_control.path().points().size() >= 2)
+            << "Empty path: " << motion_control.path().points().size() << std::endl;
+        auto destination = motion_control.path().points().at(1);
+
+        // Max distance which the robot can travel in one time step + scaling
+        // TODO (#2370): This constant is calculated multiple times.
+        float path_radius = (max_speed_ * simulator_->getTimeStep()) / 2;
+        auto path_points  = {PathPoint(
+            Vector(destination.x_meters(), destination.y_meters()), speed_at_dest)};
+        path              = AgentPath(path_points, path_radius);
+
+        // Update static obstacles
+        std::set<TbotsProto::MotionConstraint> motion_constraints;
+        for (int constraint_int : motion_control.motion_constraints())
+        {
+            if (!TbotsProto::MotionConstraint_IsValid(constraint_int))
+            {
+                continue;
+            }
+
+            const auto constraint =
+                static_cast<TbotsProto::MotionConstraint>(constraint_int);
+            auto new_obstacles =
+                obstacle_factory.createFromMotionConstraint(constraint, world);
+            if (constraint == TbotsProto::MotionConstraint::HALF_METER_AROUND_BALL)
+            {
+                ball_obstacle = new_obstacles[0];
+            }
+            else
```

Comment:
That's correct. The current implementation also considers `AVOID_BALL_PLACEMENT_INTERFERENCE` as a static obstacle. This was mainly due to `HALF_METER_AROUND_BALL` being a circle and allowing for an easy way of calculating a velocity obstacle which allows the robot to leave the obstacle if it's inside of it (Basically don't drive towards the origin of the circle). For `AVOID_BALL_PLACEMENT_INTERFERENCE` it gets a bit more complicated to decide which way to drive to leave the polygon/rectangle, so right now we just treat it the same way we treat static obstacles, which is if the robot is inside of the obstacle, then have HRVO ignore the obstacle and just follow the path calculated by the path planner out of the obstacle.

Reply:
Another way to approach this could possibly be to have the AI recognize the existent of the `HALF_METER_AROUND_BALL` obstacle, and the AI calculates the ideal place (i.e. destination) to be __outside__ of the circle. This shouldn't be too difficult and will likely lead to better results. Mainly because if the destination is in the circular obstacle, HRVO currently just tries to reach the destination while staying out of the obstacle, but it doesn't have a solid method to calculate the best place to be outside of that obstacle.